docs/software/git-software-development.md

```diff
@@ -163,3 +151,137 @@ A release of a software is created from your local clone of a software project.
+
+Another example that occasionally comes up is when you want to copy or move
+a file AND make changes to that file.
+If you have a single commit that introduces a file to a new location
+_with changes_, it will not be obvious from the commit diff itself which
+parts are the same (moved or copied in) and which parts you are modifying.
+Instead, by putting the copy or move of the original file into its own commit,
+and then putting your changes in a separate commit, it will make it clear to
+the reviewer which parts are changing from the original.
+
+#### Squash noisy work-in-progress commits
+
+Naturally in the trial-and-error-prone process of development, there will be
+many changes along the way that didn't make the final cut.
+This is good and healthy, and it's perfectly reasonable to be making many
+small work-in-progress commits locally while you are developing.
+
+However, for the reviewer, the relevant thing is what the final changes are
+being introduced into the codebase.
+Having to review several ideas that were put in to and then taken out of the
+changeset is a distraction, and makes it harder to see what the end result is
+for the new changes.
+
+If you have such work-in-progress commits, first combine them (this is also
+called "squashing" or "rebasing"), and then break them up into logically
+distinct commits as necessary, representing the final changeset.
+As mentioned above, one way to do this is with a combination of `git rebase -i`
+and `git gui`, though there are other third-party tools available also.
+
+#### Write a succinct subject to explain what each commit does
+
+The first line of a commit message is the "subject" (or sometimes also called
+the "title").
+It should be short and sweet (at most 72 characters) and briefly state _what_
+the commit is designed to do.
+
+As a convention, the subject of the commit message should be written in the
+imperative - that is, it should be written as if it were a command.
+For instance, a subject should start with "fix a bug" rather than "fixing" or
+"fixes".
+
+#### Explain why a change was made in the commit message
+
+It is generally important also to explain _why_ a change was made.
+
+If this is not covered by the succinct subject line of your commit, you should
+explain the rationale behind your change in the commit message body.
+(The commit message starts with the subject line, then is optionally followed
+by a blank line plus the message body.)
+
+You can also explain in the commit message body _how_ this commit accomplishes
+the stated purpose in the subject - and you may find yourself needing to do
+this if there are some tricky details in the changes.
+But even if it is perfectly clear from the code and your commit message _what_
+you are changing and _how_ you are going about it, it is not always clear
+_why_ the change is needed or desired - so it is important to explain your
+reasons, in order to make this clear to the reviewer.
+
```

Comment:
An example would be nice. Not to toot my own horn but I'm proud of this commit body https://github.com/opensciencegrid/topology/commit/c3524138ac8d46eee2a3c33cb75fac50acab41c4

Reply:
Sure yeah we can use that one - updated.

...

Btw speaking of tooting my own horn, i had a quick skim of the multi-line commit messages in

```
~/work/software/technology$ git log --author=edquist --no-merges | less
```

and i was not disappointed ;)